File: python/pecos/simulators/cuquantum/cuquantum_wrapper/examples/bell_state.cpp
// Copyright 2022 The PECOS developers
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
// the License.You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

// Initial author: Tyson Lawrence

#include <iostream>

#include "custatevec_workspace.hpp"
#include "gate.hpp"
#include "state_vector.hpp"
#include "utils.hpp"
#include "quantum_volume.hpp"
#include "version.h"

#include "Eigen/KroneckerProduct"

using Eigen::KroneckerProduct;

// macro to print simple things to stdout
#define PRINT(x) std::cout << x << std::endl;
#define ELAPSED(timer) std::cout << "Elapsed: " << timer.elapsed() << " s" << std::endl;

void bell_state(double randnum=0)
{
    // Initialize the workspace
    CuStatevecWorkspace workspace;

    StateVector sv = create_zero_state_vector(2);

    PRINT("Initial State");
    sv.print();

    sv.copy_to_device();

    // Create and apply the gates
    Gate h = Hadamard();
    Gate cnot = CNOT();

    h.copy_to_device();
    cnot.copy_to_device();

    h.apply(sv, workspace, {}, {0});
    cnot.apply(sv, workspace, {0}, {1});

    sv.read_from_device();

    PRINT("Bell state");
    sv.print();

    BasisBits basis_bits = {0,1};
    int parity0 = -1;
    int parity1 = -1;
    sv.measure(workspace, parity0, {0}, randnum);
    sv.measure(workspace, parity1, {1}, randnum);

    sv.read_from_device();

    PRINT("Measurement");
    std::cout << "Random Number: " << randnum << std::endl;
    std::cout << "        Bit 0: " << parity0 << std::endl;
    std::cout << "        Bit 1: " << parity1 << std::endl;
    PRINT("Final State Vector");
    sv.print();
}


int main(int argc, char *argv[])
{
    // bell_state();

    double randnum = 0;
    if (argc > 1)
        randnum = std::atof(argv[1]);

    if (randnum < 0)
        randnum = 0;

    if (randnum >= 1)
        randnum = 0.9999;

    bell_state(randnum);

    return 0;
}

File: exp/pecos-zx/examples/bell_state.rs
// Copyright 2026 The PECOS Developers
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Bell state circuit -> ZX -> simplify -> render all formats
//!
//! Demonstrates the core pipeline:
//! 1. Build a Bell state circuit in PECOS
//! 2. Convert to a ZX graph
//! 3. Render the ZX graph (ASCII + SVG + TikZ)
//! 4. Apply Clifford simplification
//! 5. Render the simplified graph

use pecos_quantum::DagCircuit;
use pecos_zx::convert::dag_to_zx;
use pecos_zx::graph::graph_stats;
use pecos_zx::simplify;
use pecos_zx::viz::Renderer;

fn main() {
    // 1. Build a Bell state circuit: H(0), CX(0,1)
    let mut dag = DagCircuit::new();
    dag.h(0);
    dag.cx(0, 1);

    println!("=== Bell State Circuit -> ZX ===\n");

    // 2. Convert to ZX graph
    let mut graph = dag_to_zx(&dag).expect("conversion failed");

    let stats = graph_stats(&graph);
    println!("Before simplification:");
    println!(
        "  Vertices: {} ({} Z-spiders, {} X-spiders, {} boundaries)",
        stats.num_vertices, stats.num_z_spiders, stats.num_x_spiders, stats.num_boundaries
    );
    println!("  Edges: {}", stats.num_edges);
    println!(
        "  I/O: {} inputs, {} outputs\n",
        stats.num_inputs, stats.num_outputs
    );

    // 3. Render before simplification
    let mut r = Renderer::default();
    r.set_output_dir("exp/pecos-zx/examples/output");
    r.render(&graph, "bell_before");

    // 4. Apply Clifford simplification
    simplify::clifford_simp(&mut graph);

    let stats_after = graph_stats(&graph);
    println!("\nAfter clifford_simp:");
    println!(
        "  Vertices: {} ({} Z-spiders, {} X-spiders, {} boundaries)",
        stats_after.num_vertices,
        stats_after.num_z_spiders,
        stats_after.num_x_spiders,
        stats_after.num_boundaries
    );
    println!("  Edges: {}\n", stats_after.num_edges);

    // 5. Render after simplification
    r.render(&graph, "bell_after");
}
